src/new_api_v04/constraint.rs

```diff
@@ -0,0 +1,116 @@
+//! A constraint is a triple (subj `pred obj).
+//!
+//! Subject and object may be variables or values.
+//!
+//! The predicate is either an AssignPredicate or a FilterPredicate.
+
+use super::{
+    predicate::{AssignPredicate, FilterPredicate, Predicate},
+    variable::VariableScope,
+};
+
+/// A litteral for subject and object in constraints.
+///
+/// Litterals are either a value from the predicate universe, or a variable
+/// that will be bound at runtime to a value in the universe.
+pub enum ConstraintLitteral<V, U> {
+    Variable(V),
+    Value(U),
+}
+
+impl<V, U> ConstraintLitteral<V, U> {
+    /// Evaluate a litteral to a value in U.
+    ///
+    /// If the litteral is a value, unwrap it. Otherwise, use the variable scope
+    /// to resolve the binding.
+    ///
+    /// If the variable is not defined, this will panic.
+    pub fn evaluate<'a>(&'a self, scope: &'a impl VariableScope<V, U>) -> &'a U {
+        match &self {
+            ConstraintLitteral::Variable(var) => scope.get(var).expect("Unbound variable"),
+            ConstraintLitteral::Value(val) => val,
+        }
+    }
+}
+
+/// A constraint for pattern matching.
+///
+/// Constraints must be carefully ordered when being satisfied as they make
+/// assumptions on which variable bindings exist:
+///  - For AssignPredicates, the lhs, if it is a variable, must be already bound,
+///    and the rhs must always a variable.
+///  - For FilterPredicates, both lhs and rhs must be bound if they are variables.
+pub struct Constraint<V, U, AP, FP> {
+    lhs: ConstraintLitteral<V, U>,
+    predicate: Predicate<AP, FP>,
+    rhs: ConstraintLitteral<V, U>,
```

Comment:
`lhs` and `rhs` do not necessarily share the same universe.

That also affects `Predicate`; we should be able to express `<node1> 'hasArity 4`.

If you want to avoid the variable soup of type variables you could add the `AP: AssignPredicate, FP: ...` constrait to the type.